src/target/devof12e/tw8816.c

```diff
@@ -29,19 +31,230 @@
+  i2c_send_start(I2C1);
+
+  /* Wait for master mode selected */
+  timeout = FLAG_TIMEOUT;
+  while(!((I2C_SR1(I2C1) & I2C_SR1_SB) & (I2C_SR2(I2C1) & (I2C_SR2_MSL | I2C_SR2_BUSY))))
+  {
+    if((timeout--) == 0) {printf("ReadErr2\n"); return i2c1_timeout();}
+  }
+
+  /* Send EEPROM address for write */
+  i2c_send_7bit_address(I2C1, deviceId, I2C_WRITE);
+
+  /* Waiting for address is transferred */
+  timeout = FLAG_TIMEOUT;
+  while(!(I2C_SR1(I2C1) & I2C_SR1_ADDR))
+  {
+    if((timeout--) == 0) {printf("ReadErr3\n"); return i2c1_timeout();}
+  }
+
+  /* Cleaning ADDR condition sequence */
+  (void)I2C_SR2(I2C1);
+
+  /* Send the EEPROM's internal address to read from: Only one byte address */
+  i2c_send_data(I2C1, periphMemAddr);
+
+  /* Waiting for byte transfer finished */
+  timeout = FLAG_TIMEOUT;
+  while (!(I2C_SR1(I2C1) & (I2C_SR1_BTF | I2C_SR1_TxE)))
+  {
+    if((timeout--) == 0) {printf("ReadErr4\n"); return i2c1_timeout();}
+  }
+
+  /* Send START condition a second time */
+  i2c_send_start(I2C1);
+
+  /* Wait for master mode selected */
+  timeout = FLAG_TIMEOUT;
+  while(!((I2C_SR1(I2C1) & I2C_SR1_SB) & (I2C_SR2(I2C1) & (I2C_SR2_MSL | I2C_SR2_BUSY))))
+  {
+    if((timeout--) == 0) {printf("ReadErr5\n"); return i2c1_timeout();}
+  }
+
+  /* Send EEPROM address for read */
+  i2c_send_7bit_address(I2C1, deviceId, I2C_READ);
+
+  /* Waiting for address is transferred */
+  timeout = FLAG_TIMEOUT;
+  while(!(I2C_SR1(I2C1) & I2C_SR1_ADDR))
+  {
+    if((timeout--) == 0) {printf("ReadErr6a\n"); return i2c1_timeout();}
+  }
+
+  /* Cleaning ADDR condition sequence */
+  (void)I2C_SR2(I2C1);
+
+  /* Enable Acknowledgement to be ready for another reception */
+  i2c_enable_ack(I2C1);
+
+  /* While there is data to be read */
+  while(len) {
+      if(len == 1) {
+          /* Disable Acknowledgement */
+          i2c_disable_ack(I2C1);
+
+          /* Send STOP Condition */
+          i2c_send_stop(I2C1);
+      }
+
+      /* Wait for the byte to be received */
+      timeout = FLAG_TIMEOUT;
+      while(!(I2C_SR1(I2C1) & (I2C_SR1_RxNE)))
+      {
+        if((timeout--) == 0) {printf("ReadErr6b\n"); return i2c1_timeout(); }
+      }
+
+      /* Read a next byte */
+      *buffer = i2c_get_data(I2C1);
+
+      /* Point to the next location where the byte read will be saved */
+      buffer++;
+
+      /* Decrement the read bytes counter */
+      len--;
+  } //while(len)
+
+  /* Wait to make sure that STOP control bit has been cleared */
+  timeout = FLAG_TIMEOUT;
+  while(I2C_CR1(I2C1) & I2C_CR1_STOP)
+  {
+    if((timeout--) == 0) {printf("ReadErr6c\n"); return i2c1_timeout();}
+  }
+
+  /* Re-Enable Acknowledgement to be ready for another reception */
+  i2c_enable_ack(I2C1);
+
+  /* If all operations OK, return sEE_OK (0) */
+  return 0;
+}
+
+static u32 LCD_ReadReg(unsigned reg)
+{
+    u8 val;
+    for (int i = 0; i < 2; i++) {
+        if (! I2C1_ReadBuffer(0x45, &val, reg, 1))
+            break;
+    }
+    return val;
```

Comment:
We still get a compiler warning here about `val` possibly being uninitialized

Reply:
i will fix it in #503

Reply:
already fixed in master.